### PR TITLE
Feature/townhall UI rework

### LINKED ISSUE #9558: [BUG] Courier AI gets bugged when trying to access the warehouse
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using the latest MineColonies Version?

- [X] I am running the latest **alpha** version of MineColonies for my Minecraft version.

### Did you check on the Wiki? or ask on Discord?

- [X] I checked the MineColonies Wiki and made sure my issue is not covered there. Or I was sent from discord to open an issue here.

### What were you playing at the time? Were you able to reproduce it in both settings?

- [ ] Single Player
- [X] Multi Player

### Minecraft Version

1.19.2

### MineColonies Version

1.1.304

### Structurize Version

1.0.661

### Related Mods and their Versions

_No response_

### Current Behavior

Courier AI often bugs out when they try to access the warehouse
image

### Expected Behavior

Couriers working as usual

### Reproduction Steps

1. Create a colony with a warehouse and courier hut
2. Use the couriers to deposit items into the warehouse or deliver items from the warehouse

*This is the best I can do because it seems to happen randomly as far as I'm aware. Log files may indicate as to when this could happen but I have no clue what they mean.

### Logs

https://gist.github.com/xlightningstar/6c822688d02dbe471a5352e13acd60c5

### Anything else?

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #9572: [BUG] [1.19.2] Citizens constantly jump in and out of bed when raining.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using the latest MineColonies Version?

- [X] I am running the latest **alpha** version of MineColonies for my Minecraft version.

### Did you check on the Wiki? or ask on Discord?

- [X] I checked the MineColonies Wiki and made sure my issue is not covered there. Or I was sent from discord to open an issue here.

### What were you playing at the time? Were you able to reproduce it in both settings?

- [ ] Single Player
- [X] Multi Player

### Minecraft Version

1.19.2

### MineColonies Version

1.19.2-1.1.310-BETA

### Structurize Version

1.19.2-1.0.649-BETA

### Related Mods and their Versions

Forge: 43.2.21

### Current Behavior

When it is raining, citizens are tasked with sleep as their desired activity, but since it is daytime, they get right back out of bed. So they keep jumping in and out of bed the entire time.

### Expected Behavior

Stay in bed.

Or more ideally, just stay at home/the tavern/restaurant/leisure sites/other suitable locations, and wander around.

### Reproduction Steps

Have citizens
Have houses
Have rain
Observe

### Logs

N/A

### Anything else?

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #9576: [BUG] Warehouse only highlighting one set of searched objects at a time
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using the latest MineColonies Version?

- [X] I am running the latest **alpha** version of MineColonies for my Minecraft version.

### Did you check on the Wiki? or ask on Discord?

- [X] I checked the MineColonies Wiki and made sure my issue is not covered there. Or I was sent from discord to open an issue here.

### What were you playing at the time? Were you able to reproduce it in both settings?

- [ ] Single Player
- [X] Multi Player

### Minecraft Version

1.20.1

### MineColonies Version

1.1.308-BETA

### Structurize Version

1.0.662-BETA

### Related Mods and their Versions

Forge 47.1.3
BlockUI 1.0.119-BETA.jar

### Current Behavior

If I search in the warehouse for, say, cobblestone, it tells me there is 4k or so. Then it only highlights one small stack. So, I go get that and have to search again. The next search does the same. It seems to show the smallest stacks first, and only one at a time. 

This was also happening on the Raid server too (from version 1.1.274-BETA), so I don't think it's my mod pack (ATM9 right now).

### Expected Behavior

The interface used to highlight ALL the searched item across all racks.

### Reproduction Steps

Have an unsorted warehouse.
Search the inventory for an item that is scattered throughout.
Only one rack will highlight. If you remove those items and search again, only one different rack will be highlighted. It never highlights multiple racks anymore.

### Logs

https://gist.github.com/kezmodius/5cb4999b3ee642a355c9d60f3a9ff517

### LINKED ISSUE #9577: [BUG] Guards keeps requesting armor, even though they have some already.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using the latest MineColonies Version?

- [X] I am running the latest **alpha** version of MineColonies for my Minecraft version.

### Did you check on the Wiki? or ask on Discord?

- [X] I checked the MineColonies Wiki and made sure my issue is not covered there. Or I was sent from discord to open an issue here.

### What were you playing at the time? Were you able to reproduce it in both settings?

- [ ] Single Player
- [X] Multi Player

### Minecraft Version

1.20.1

### MineColonies Version

1.20.1-1.1.314-BETA

### Structurize Version

1.20.1-1.0.662-BETA

### Related Mods and their Versions

_No response_

### Current Behavior

After fulfilling the guard's request for armor, a few moments later they request armor again. 
image
image

### Expected Behavior

Once they have armor, they should request more.

### Reproduction Steps

fulfill their armor
Wait
Check their requests. 

### Logs

N/A

### Anything else?

I have not tried on the newest available version, but was told on the discord to submit an issue anyways. 

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### LINKED ISSUE #9578: [BUG] Colony won't fill archer or guard slots
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using the latest MineColonies Version?

- [X] I am running the latest **alpha** version of MineColonies for my Minecraft version.

### Did you check on the Wiki? or ask on Discord?

- [X] I checked the MineColonies Wiki and made sure my issue is not covered there. Or I was sent from discord to open an issue here.

### What were you playing at the time? Were you able to reproduce it in both settings?

- [X] Single Player
- [ ] Multi Player

### Minecraft Version

1.19.2

### MineColonies Version

1.1.275-beta

### Structurize Version

1.0.660

### Related Mods and their Versions

_No response_

### Current Behavior

I have unemployed colonists (5 in the screenshot, now 9). I have 0 archers or soldiers. The druid role is filled though, it seems to the the only 'guard' role that actually auto-hires. When I built the stone smeltery, that job was filled immediately also. 

### Expected Behavior

For all guard types to auto-hire new citizens when guards die or new towers are built, not just druids.

### Reproduction Steps

1. Have unemployed citizens
2. Build guard towers or have guards die.
3. Druids are the only guard type that will refill

### Logs

https://gist.github.com/purpletiger6862/924b696b2339f078a69d14eb7d2490f9.js

### Anything else?

* Add a thumbs-up to the bug report if you are also affected. This helps the bug report become more visible to the team and doesn't clutter the comments.
* Add a comment if you have any insights or background information that isn't already part of the conversation.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #9577
Closes #9576
Closes #9578
Closes #9558
Closes #9572

# Changes proposed in this pull request:
- Move settings to main page
- Move happiness to citizen page, rework citizen page and displayed info
- Adjust happiness icons
- Rework main page with new icons
- Move permission events to permission page
- Move alliance to its own page
- Update views on newsub registration already
- Move normal colony events together with workorders (one page less now)
- Make all jobs in job stat list upper case first letter.
- cleanup permission page

- Fix an issue with highlightmanager
- Fix issue with guard requests

Sorry for the Omnibus:

- When raining, citizens go into wander AI and not sleeping AI. May go to library, townhall, tavern or home.

[ x ] Yes I tested this before submitting it.
[ ] I also did a multiplayer test.

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/client/gui/AbstractWindowModuleBuilding.java [hunk 1]
```diff
@@ -60,6 +59,13 @@ public AbstractWindowModuleBuilding(final B building, final String resource)
         }
     }
 
+    @Override
+    public void setPage(final boolean relative, final int page)
+    {
+        super.setPage(relative, page);
+        mc.getSoundManager().play(SimpleSoundInstance.forUI(SoundEvents.BOOK_PAGE_TURN, 1.0F));
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/AbstractWindowModuleBuilding.java:66]: Move sound to super would make more sense?

File: src/main/java/com/minecolonies/coremod/client/gui/AbstractWindowSkeleton.java [hunk 2]
```diff
@@ -73,6 +74,9 @@ public AbstractWindowSkeleton(final String resource, @Nullable final BOWindow pa
         {
             buttonNextPage = findPaneOfTypeByID(BUTTON_NEXTPAGE, Button.class);
             buttonPrevPage = findPaneOfTypeByID(BUTTON_PREVPAGE, Button.class);
+            PaneBuilders.tooltipBuilder().hoverPane(buttonNextPage).build().setText(Component.translatable("com.minecolonies.core.gui.nextpage"));
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/AbstractWindowSkeleton.java]: Fyi: there is single line tooltip shortener in PaneBuilders

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowCitizenPage.java [hunk 3]
```diff
@@ -114,7 +253,29 @@ public int getElementCount()
...
+                final AbstractTextBuilder.TextBuilder textBuilder = PaneBuilders.textBuilder();
+                for (final Map.Entry<Skill, Tuple<Integer, Double>> entry : citizen.getCitizenSkillHandler().getSkills().entrySet())
+                {
+                    final String skillName = entry.getKey().name().toLowerCase(Locale.US);
+                    final int skillLevel = entry.getValue().getA();
+
+                    textBuilder.append(Component.translatable("com.minecolonies.coremod.gui.citizen.skills." + skillName));
+                    textBuilder.append(Component.literal(": " + skillLevel + " "));
+                }
+                textBuilder.newLine();
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowCitizenPage.java]: You don't need to call this before builder.build()

**Comment by Thodor12** [comment]
Could you add some images of the final product per page?

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowInfoPage.java [hunk 4]
```diff
@@ -70,227 +61,194 @@ public WindowInfoPage(final BuildingTownHall.View building)
...
-                    int workers = module.getAssignedCitizens().size();
-
-                    final String jobName = module.getJobDisplayName().toLowerCase(Locale.ENGLISH);
+                final Text nameLabel = rowPane.findPaneOfTypeByID(NAME_LABEL, Text.class);
+                final Text actionLabel = rowPane.findPaneOfTypeByID(ACTION_LABEL, Text.class);
 
-                    final Tuple<Integer, Integer> tuple = jobMaxCountMap.getOrDefault(jobName, new Tuple<>(0, 0));
-                    jobMaxCountMap.put(jobName, new Tuple<>(tuple.getA() + workers, tuple.getB() + max));
+                final List<IColonyEventDescription> colonyEvents = building.getColonyEvents();
+                Collections.reverse(colonyEvents);
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowInfoPage.java]: Can we avoid reversing this list every row update?

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowInfoPage.java [hunk 5]
```diff
@@ -70,227 +61,194 @@ public WindowInfoPage(final BuildingTownHall.View building)
...
+                if (index == 0)
                 {
-                    final List<IColonyEventDescription> colonyEvents = building.getColonyEvents();
-                    Collections.reverse(colonyEvents);
-                    final IColonyEventDescription event = colonyEvents.get(index);
-                    if (event instanceof CitizenDiedEvent)
+                    if (numElements == 1)
                     {
-                        actionLabel.setText(Component.literal(((CitizenDiedEvent) event).getDeathCause()));
+                        rowPane.findPaneOfTypeByID(BUTTON_DOWN, Button.class).hide();
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowInfoPage.java]: setVisible(condition in if)

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java [hunk 6]
```diff
@@ -64,111 +101,376 @@ public WindowMainPage(final BuildingTownHall.View building)
...
+     * @param button the pressed button.
      */
-    private void patreonClicked()
+    private void toggleHiring(@NotNull final Button button)
     {
-        Minecraft.getInstance().setScreen(new ConfirmLinkScreen((check) -> {
-            if (check) {
-                Util.getPlatform().openUri("https://www.patreon.com/Minecolonies");
-            }
+        String key = button.getText().getContents() instanceof TranslatableContents ? ((TranslatableContents) button.getText().getContents()).getKey() : button.getTextAsString();
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java]: Instanceof pattern matching
```suggestion
        String key = button.getText().getContents() instanceof final TranslatableContents contents ? contents.getKey() : button.getTextAsString();
```

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java [hunk 7]
```diff
@@ -64,111 +101,376 @@ public WindowMainPage(final BuildingTownHall.View building)
...
      *
-     * @param button the clicked button.
+     * @param button the pressed button.
      */
-    private void teleportToColony(@NotNull final Button button)
+    private void toggleHousing(@NotNull final Button button)
     {
-        final int row = alliesList.getListElementIndexByPane(button);
-        final CompactColonyReference ally = building.getColony().getAllies().get(row);
+        String key = button.getText().getContents() instanceof TranslatableContents ? ((TranslatableContents) button.getText().getContents()).getKey() : button.getTextAsString();
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java]: Instanceof pattern matching

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java [hunk 8]
```diff
@@ -64,111 +101,376 @@ public WindowMainPage(final BuildingTownHall.View building)
...
+     *
+     * @param button the pressed button.
      */
-    @Override
-    public void onOpened()
+    private void toggleMoveIn(@NotNull final Button button)
     {
-        super.onOpened();
-        fillAlliesAndFeudsList();
+        String key = button.getText().getContents() instanceof TranslatableContents ? ((TranslatableContents) button.getText().getContents()).getKey() : button.getTextAsString();
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java]: Instanceof...

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java [hunk 9]
```diff
@@ -64,111 +101,376 @@ public WindowMainPage(final BuildingTownHall.View building)
...
-        if (building.getColony().getMercenaryUseTime() != 0
-              && building.getColony().getWorld().getGameTime() - building.getColony().getMercenaryUseTime() < TICKS_FOURTY_MIN)
+    /**
+     * Toggles printing progress.
+     *
+     * @param button the button to toggle.
+     */
+    private void togglePrintProgress(@NotNull final Button button)
+    {
+        String key = button.getText().getContents() instanceof TranslatableContents ? ((TranslatableContents) button.getText().getContents()).getKey() : button.getTextAsString();
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java]: Instanceof hopefully last time

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java [hunk 10]
```diff
@@ -64,111 +101,376 @@ public WindowMainPage(final BuildingTownHall.View building)
...
+    /**
+     * Switch the structure style pack.
+     */
+    private void switchPack()
+    {
+        new WindowSwitchPack(() -> {
+            building.getColony().setStructurePack(StructurePacks.selectedPack.getName());
+            Network.getNetwork().sendToServer(new ColonyStructureStyleMessage(building.getColony(), StructurePacks.selectedPack.getName()));
+            return new WindowMainPage((BuildingTownHall.View) this.building);
+        }).open();
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java:244]: You can try to use the layering system `openAsLayer`, which avoids open() machinery

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java [hunk 11]
```diff
@@ -64,111 +101,376 @@ public WindowMainPage(final BuildingTownHall.View building)
...
+
+    /**
+     * Opens the banner picker window. BOWindow does not use blockui, so is started manually.
+     *
+     * @param button the trigger button
+     */
+    private void openBannerPicker(@NotNull final Button button)
+    {
+        Screen window = new WindowBannerPicker(building.getColony(), this, isFeatureUnlocked);
+        Minecraft.getInstance().execute(() -> Minecraft.getInstance().setScreen(window));
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java]: If you're in ui logic u don't need execute guard

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java [hunk 12]
```diff
@@ -64,111 +101,376 @@ public WindowMainPage(final BuildingTownHall.View building)
...
+            else
+            {
+                textPane.show();
+            }
+
+            AbstractTextBuilder.TooltipBuilder hoverText = PaneBuilders.tooltipBuilder().hoverPane(textPane);
+            hoverText.append(Component.translatable("com.minecolonies.core.townhall.patreon.textures")).paragraphBreak();
+            hoverText.appendNL(Component.empty());
+            hoverText.appendNL(Component.translatable("com.minecolonies.core.townhall.patreon")).paragraphBreak();
+            hoverText.build();
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java]: It's builder so you can chain method calls :D
No need to paragraph break before build (few lines below too)

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowPermissionsPage.java [hunk 13]
```diff
@@ -266,18 +259,40 @@ public void onOpened()
...
+            playerNameField.setEnabled(true);
+            rankNameField.setEnabled(true);
+            addRankButton.setEnabled(true);
         }
         else
         {
-            label.show();
-            button.setEnabled(false);
+            AbstractTextBuilder.TooltipBuilder hoverText = PaneBuilders.tooltipBuilder().hoverPane(playerNameField);
+            hoverText.append(Component.translatable("com.minecolonies.coremod.gui.townhall.player_permission_error")).paragraphBreak();
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowPermissionsPage.java:279]: No break needed (below too)

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowPermissionsPage.java [hunk 14]
```diff
@@ -320,6 +335,23 @@ public void updateElement(final int i, final Pane pane)
...
+        dropdown.setDataProvider(new DropDownList.DataProvider()
+        {
+            @Override
+            public int getElementCount()
+            {
+                return rankTypes.size();
+            }
+
+            @Override
+            public String getLabel(final int i)
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowPermissionsPage.java:349]: Use getLabelNew if returning components, need to deprecate the old one (or remove it finally after few yrs xd)

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowStatsPage.java [hunk 15]
```diff
@@ -64,70 +70,148 @@ public WindowStatsPage(final BuildingTownHall.View townHall)
...
+                        }
+                    }
+                    int max = module.getMaxInhabitants() - alreadyAssigned + module.getAssignedCitizens().size();
+                    int workers = module.getAssignedCitizens().size();
 
+                    final String jobName = module.getJobDisplayName().toLowerCase(Locale.ENGLISH);
 
-        int yPos = 60;
-        for (final Map.Entry<String, Double> entry : happinessMap.entrySet())
+                    final com.minecolonies.api.util.Tuple<Integer, Integer> tuple = jobMaxCountMap.getOrDefault(jobName, new com.minecolonies.api.util.Tuple<>(0, 0));
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowStatsPage.java]: FQN import?

**Comment by Thodor12** [comment]
What if we still strip the "statistics" tab, in favor of the information tab?

File: src/main/resources/assets/minecolonies/gui/townhall/windowtownhall.xml [hunk 16]
```diff
@@ -2,71 +2,61 @@
 
     <image source="minecolonies:textures/gui/townhall_book.png" pos="75 0" size="374 243"/>
```
Comment (Nightenom) [inline src/main/resources/assets/minecolonies/gui/townhall/windowtownhall.xml:4]: We have six ribbon textures, can you use them all equally instead one many times?
Plus you need to check extended versions down matches the IDs here

**Comment by Raycoms** [comment]
> What if we still strip the "statistics" tab, in favor of the information tab?

I was thinking about this, but I was wanting to avoid to have to find this info by clicking through pages

**Comment by Thodor12** [comment]
Ok and one more thing, in the code I noticed a "WindowAlliancePage", where actually did that go, is that the last page you did not show off in the images yet?

**Comment by Raycoms** [comment]
> Ok and one more thing, in the code I noticed a "WindowAlliancePage", where actually did that go, is that the last page you did not show off in the images yet?

Didnt show it because it doesnt look interesting xD Nothing changes, it just has the Allies/feuds from the main page

**Comment by Thodor12** [comment]
Fair enough, but that is the last page in the list then, good to know

**Review by Nightenom — CHANGES_REQUESTED** [review]
Most of them are nitpicks - apply as you want, some of them across entire PR, few actually wanted

PaneBuilder for text apply paragraphBreak in build method

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowPermissionsPage.java (hunk 14 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowPermissionsPage.java:349]: getLabel is required

File: src/main/java/com/minecolonies/coremod/client/gui/AbstractWindowModuleBuilding.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/client/gui/AbstractWindowModuleBuilding.java:66]: I didnt put that there intentionally, as I didnt want it to apply to ALL uis, only to those specific ones.

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowPermissionsPage.java (hunk 14 above)
Comment (Nightenom) [reply src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowPermissionsPage.java:349]: Tru, but not called if you override component variant

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java [hunk 17]
```diff
@@ -175,9 +176,20 @@ private IState calculateNextState()
         }
         else
         {
-            if (citizen.getCitizenSleepHandler().isAsleep() && !citizen.getCitizenDiseaseHandler().isSick())
+            if (citizen.getCitizenSleepHandler().isAsleep())
             {
-                citizen.getCitizenSleepHandler().onWakeUp();
+                if (citizen.getCitizenDiseaseHandler().isSick())
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/CitizenAI.java:181]: odd change, before we only did wakeup if not sick, now we do if sick?

Also whats with the null and distance checks before calling it, why do they apply to sick only?

**Review by someaddons — APPROVED** [review]

**Pull request merged**